src/mpi/coll/op/opequal.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpiimpl.h"
+
+/* Equal structures */
+/* For reusing code, using int type instead of bool*/
+typedef struct MPIR_2int_eqltype {
+    int value;
+    int is_equal;
+} MPIR_2int_eqltype;
+
+typedef struct MPIR_longint_eqltype {
+    long value;
+    int is_equal;
+} MPIR_longint_eqltype;
+
+typedef struct MPIR_shortint_eqltype {
+    short value;
+    int is_equal;
+} MPIR_shortint_eqltype;
+
+
+/* If a child found unequal, its parent sticks to unequal.*/
+/* Values of is_equal: 1 equal 0 not equal, init using any value other than 0.*/
+#define MPIR_EQUAL_C_CASE_INT(c_type_, _len_) {             \
+        c_type_ *a = (c_type_ *)inoutvec;                   \
+        c_type_ *b = (c_type_ *)invec;                      \
+        for (i = 0; i < _len_; ++i) {                       \
+            if (!(b[i].is_equal) || !(a[i].is_equal)){      \
+                a[i].is_equal = 0;                          \
+            }else if (a[i].value != b[i].value){            \
+                a[i].is_equal = 0;                          \
+            }else{                                          \
+                a[i].is_equal = 1;                          \
+            }                                               \
+        }                                                   \
+    }                                                   
+
+
+void MPIR_EQUAL_struct(void *invec, void *inoutvec, int *Len, MPI_Datatype * type)
+{
+    int len = *Len, data_len = 0, type_len = 0;
+    int i, j, is_equal, *ints;
+    void *invec_i_pos, *invec_i_bool_pos, *inoutvec_i_pos, *inoutvec_i_bool_pos;
+    MPI_Aint *aints, *counts, lb, extent, size;
+    MPI_Datatype *types;
+    MPIR_Datatype *typeptr;
+    MPIR_Datatype_contents *cp;
+
+    /* decode */
+    MPIR_Datatype_get_ptr(*type, typeptr);
+
+    cp = typeptr->contents;
+
+    MPIR_Datatype_access_contents(cp, &ints, &aints, &counts, &types);
+
+    MPIR_Type_get_extent_impl(*type, &lb, &extent);
+    type_len = extent - lb;
+
+    /*Compare structs*/
+    for (i = 0; i < len; ++i)       
+    {
+        invec_i_pos = invec + i*type_len;
+        invec_i_bool_pos = invec_i_pos + aints[cp->nr_aints-1];
+        inoutvec_i_pos = inoutvec + i*type_len;
+        inoutvec_i_bool_pos = inoutvec_i_pos + aints[cp->nr_aints-1];
+
+        if(!(*(int*)invec_i_bool_pos) ||
+            !(*(int*)inoutvec_i_bool_pos))
+        {/*compare values of is_equal*/
+            *(int*)inoutvec_i_bool_pos  = 0;
+        }
+        else
+        {/*compare the content of the struct*/
+            is_equal = 1;
+
+            for(j = 0; j < cp->nr_types-1; ++j)
+            {
+                MPIR_Type_size_impl(types[j], &size); 
+                data_len = ints[j + 1] * size;
+
+                if(memcmp(invec_i_pos+aints[j], inoutvec_i_pos+aints[j], data_len))
```

Comment:
Code like this is difficult to read, use some local variables with good names to make the code easier to read